### ISSUE TITLE
Replace usage of file with open in twisted.python.versions

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5785]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#5785|
|Type|task|
|Created|2012-07-15 19:00:26Z|
|Branch|https://github.com/twisted/twisted/tree/versions-file-5785|

`twisted.python.versions` contains 2 `file` calls that are deprecated in Python 3, and should be replaced with `open` instead.

```
twisted/python/versions.py:212:                format = file(formatFile).read().strip()
twisted/python/versions.py:223:            entries = file(entriesFile)
```

Attached patch replaces these calls and also brings code coverage from 92 to 98%.

Attachments:

* [versions-5785.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1d7/1d7a6457332c8a49978a5db3aca4b36d0c62fec6/bab969a8f19a165ab672d04c588b1f92e4c56666.patch) (4535 bytes) - added by thijs on 2012-07-15 19:01:02Z - 
* [versions-5785-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1d7/1d7a6457332c8a49978a5db3aca4b36d0c62fec6/04df61f6045e17a99ebc051c6288d0d7b8f22a2d.patch) (6099 bytes) - added by thijs on 2012-07-16 09:26:49Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5785 5785
type__task task
reporter__thijs thijs
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_versions_file_5785 branches/versions-file-5785
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1342378826000000 1342378826000000
changetime__1346241741000000 1346241741000000
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_2"></a>|@pitrou commented|
|-|-|

I can't comment on the new tests, but the change in twisted.python.versions looks obviously fine.
Of course, you could have taken the opportunity to use a "with" statement (and perhaps made explicit the "r" argument to open(), but that's a personal preference) :-)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Replying to [antoine](#note_2):
> I can't comment on the new tests, but the change in twisted.python.versions looks obviously fine.
> Of course, you could have taken the opportunity to use a "with" statement (and perhaps made explicit the "r" argument to open(), but that's a personal preference) :-)

Thanks for the review, I added the 'r' argument in the latest patch.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_4"></a>|@vperic set owner to @thijstriemstra|
|-|-|

This looks good to me. There's also some overlap with ticket [#5627](https://github.com/twisted/twisted/issues/5627), but lets wait for this to get in and I'll take care of the other ticket. I guess a topfile should be added and tests run, but other than that +1 from me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [34948]) Apply versions-5785-2.patch: Replace the usage of file with open in twisted.python.versions for Python 3 compatibility.

Author: thijs
Reviewer: antoine, vperic
Fixes: [#5785](https://github.com/twisted/twisted/issues/5785)
Refs [#5627](https://github.com/twisted/twisted/issues/5627)

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_6"></a>|@vperic set status to `reopened`|
|-|-|

(In [34973]) Revert r34948: Apply versions-5785-2.patch

The commit added a test, test_dontAllowOtherObjects, which compares a
Version with a string object. It is however not reliable, as it compares
memory addresses, so revert the whole commit.

Reopens: [#5785](https://github.com/twisted/twisted/issues/5785)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set status to `new`|
|-|-|

Replying to [vperic](#note_6):
> (In [34973]) Revert r34948: Apply versions-5785-2.patch
> 
> The commit added a test, test_dontAllowOtherObjects, which compares a
> Version with a string object. It is however not reliable, as it compares
> memory addresses, so revert the whole commit.
> 
> Reopens: [#5785](https://github.com/twisted/twisted/issues/5785)

Ugh. I will prepare a new patch for this.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_8"></a>|@vperic commented|
|-|-|

Hi, in [#5627](https://github.com/twisted/twisted/issues/5627) I submitted a patch to remove cmp which has a similar test, so please don't duplicate work.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [vperic](#note_8):
> Hi, in [[#5627](https://github.com/twisted/twisted/issues/5627)](https://github.com/twisted/twisted/issues/5627) I submitted a patch to remove cmp which has a similar test, so please don't duplicate work.

I'm aware of this and will hold off with this work until #5627 has landed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

(In [35325]) Branching to 'versions-file-5785'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [35431]) Merge versions-python3-5917-3

Author: exarkun
Reviewer: itamar
Fixes: [#5917](https://github.com/twisted/twisted/issues/5917)
Fixes: [#5785](https://github.com/twisted/twisted/issues/5785)
Fixes: [#5886](https://github.com/twisted/twisted/issues/5886)

Port `twisted.python.versions` to Python 3.  Also add a bit of missing test coverage
and restore the definition of `twisted.__version__` on Python 3.
